add QueryAPI pages (#1479)
near/docs: docs/bos/queryapi/how-works.md

---
id: how-it-works
title: How QueryAPI works
sidebar_label: How it works
---

> QueryApi is a streaming indexer implementation that executes queries written by developers on the NEAR blockchain.

Reviewer: I'm not sure `queries` is the best term here, as it might be confused with GraphQL queries? Best I can come up with is 'indexer code' but that doesn't seem great 😅

Author: I used `... executes custom indexing logic written ...`. Feel free to edit or change it 👍

> QueryApi allows hosted execution of complex queries (ones that can’t be answered by a [simple RPC](../../5.api/rpc/introduction.md) or [Enhanced API](https://docs.pagoda.co/api) call) and hosting of the data output of those queries using GraphQL endpoints.

Reviewer: ```suggestion
QueryApi allows hosted execution of complex queries (ones that can’t be answered by a [simple RPC](../../5.api/rpc/introduction.md) or [Enhanced API](https://docs.pagoda.co/api) call), hosting of the data, and its exposure via GraphQL endpoints.
```
Technically GraphQL doesn't 'host', it exposes just exposes the data

## Components involved 

The QueryApi implementation integrates many different components in a single and streamlined solution.
In a high-level overview, the main components are:

:::info Components

> `NEAR Protocol` -> `NEAR Lake` -> `Coordinator` -> `Database` -> `API`

Reviewer: ```suggestion
`NEAR Protocol` -> `NEAR Lake` -> `Coordinator` -> `Runner` -> `Database` -> `API`
```
We might want to mention that `Runner` executes the user code which outputs the data to the database?

:::

### Detailed overview

An in-depth, detailed overview of the QueryApi components:

[![QueryAPI](/docs/qapi-components.png)](/docs/qapi-components.png)

### Description

- **Protocol:** the underlying NEAR Layer-1 Blockchain, where data `Blocks` and `Chunks` are produced.

> - **NEAR Lake:** an indexer built on top of the [NEAR Lake Framework](../../1.concepts/3.advanced/near-lake-framework.md) to watch the Layer-1 network and store all the events as JSON files on AWS S3. Changes are indexed as new `Blocks` arrive.

Reviewer: ```suggestion
- **NEAR Lake:** an indexer which watches the Layer-1 network and stores all the events as JSON files on AWS S3. Changes are indexed as new `Blocks` arrive.
```
`Near Lake Framework` is actually the libraries which consumer from Near Lake/S3

> - **Coordinator**: the QueryApi coordinator indexer filters matching data `Blocks`, runs historical processing threads, and executes developer's JS code.

Reviewer: ```suggestion
- **Coordinator**: the QueryApi coordinator indexer filters matching data `Blocks`, runs historical processing threads, and queues developer's JS code to be indexed with these matched blocks.
```
`Coordinator` doesn't execute the functions, AWS Lambda does

> - **Database**: a Postgres database where filtered and indexed results are stored, using a Logical DB per user, and a Logical schema per indexer function.

Reviewer: ```suggestion
- **Database**: a Postgres database where developer indexer data is stored, using a Logical DB per user, and a Logical schema per indexer function.
```
Not sure if this is the best wording, but Postgres stores the data which is output from developers code

> - **API**: a Hasura GraphQL server running on Google Cloud Platform exposes simple web API endpoints so users can access GraphQL queries and subscriptions from anywhere.

Reviewer: ```suggestion
- **API**: a Hasura server running on Google Cloud Platform exposes a GraphQL endpoint so users can access their data from anywhere.
```
Technically it's one (GraphQL) endpoint

## Who hosts QueryAPI

> [Pagoda Inc.](https://pagoda.co) runs and manages all the infrastructure of QueryAPI, including NEAR Lake nodes to store the data in JSON format on AWS S3.

Reviewer: I wonder if this statement will still be true after Protocol spin out, since Near Lake will be owned by them 😅

Author: We'll have to update it after the spin-off 😄

- NEAR Lake indexing is hosted on AWS S3 buckets.

> - Hasura GraphQL API server is hosted on GCP.

Reviewer: Do we want to mention that the other hosted components? Coordinator on GCP, Runners on AWS SQS & Lambda?

Author: Added
